spec/index.html

```diff
@@ -791,123 +853,695 @@ <h3>Converting a <a>YAML document</a></h3>
+      </p>
+
+      <p>
+        As with the <a>YAML-LD JSON profile</a>,
+        keys used in a <a>YAML mapping</a> MUST be <a>strings</a>.
+      </p>
+
+      <p>
+        YAML-LD docucments MAY use <a>alias nodes</a>,
+        as long as dereferencing these aliases does not result in a loop.
+      </p>
+
+      <p>
+        As with the <a>YAML-LD JSON profile</a>,
+        a <a>YAML stream</a> MUST include a single <a>YAML document</a>,
+        as the <a>JSON-LD extended internal representation</a> supports
+        only a single document model.
+      </p>
+
+      <p class="issue" data-number="79">
+        Consier something like `!id` as a local tag to denote IRIs.
+      </p>
+
+      <section id="extended-internal-representation">
+        <h4>The JSON-LD Extended Internal Representation</h4>
+
+        <p>
+          This specification defines
+          the <dfn data-lt="extended internal representation">
+            JSON-LD extended internal representation
+          </dfn>, an extension
+          of the JSON-LD <a>internal representation</a>.
+        </p>
+
+        <p>
+          In addition to <a>maps</a>, <a>arrays</a>, and <a>strings</a>,
+          the <a>internal representation</a> allows native representation
+          of <a>numbers</a>, <a>boolean</a> values, and <a>nulls</a>.
+          The <a>extended internal representation</a> allows for native
+          representation of <a>RDF literals</a>, both
+          with a <a>datatype IRI</a>,
+          and <a>language-tagged strings</a>.
+        </p>
+
+        <p>
+          When transforming from the <a>extended internal representation</a>
+          to the <a>internal representation</a> &mdash;
+          for example when serializing to JSON
+          or to the <a>YAML-LD JSON profile</a> &mdash;
+          implementations MUST transform <a>RDF literals</a> to the closest
+          native representation of the <a>internal representation</a>:
+        </p>
+
+        <ul>
+          <li>
+            Literals with datatype
+            <a data-cite="xmlschema11-2#boolean">`xsd:boolean`</a>
+            are transformed to either `true` or `false`,
+          </li>
+          <li>
+            Literals with datatype
+            <a data-cite="xmlschema11-2#decimal">`xsd:decimal`</a>
+            are transformed a native <a>number</a>,
+          </li>
+          <li>
+            Literals with datatype
+            <a data-cite="xmlschema11-2#decimal">`xsd:decimal`</a>,
+            <a data-cite="xmlschema11-2#decimal">`xsd:double`</a>,
+            <a data-cite="xmlschema11-2#decimal">`xsd:float`</a>,
+            or derived datatypes,
+            are transformed a native <a>number</a>,
+          </li>
+          <li>
+            All other literals a transformed to a native <a>string</a>.
+          </li>
+        </ul>
+
+        <p class="ednote">
+          An alternative would be to transform such literals to
+          JSON-LD <a>value objects</a>,
```

Comment:
I reallyy don't see the benefit of losing information at this stage, while the (standard) internal representation is capable of representing RDF literals through value objects...

Reply:
My concern was for round-tripping: how do you distinguish from something specified using a value object (or through context coercion) from something expressed natively? How would you know to express a value object using node tags? Easiest to not do that at all, and allow direct representation within the (expanded) internal representation. At least, this was the conclusion I came to when thinking about it over time.

Being able to represent more native values in the internal representation (also potentially IRIs and blank nodes) better supports the greater expressivity of alternative formats such as YAML and CBOR. And, it's largely transparent to the JSON-LD processing algorithms, other than the to/from RDF bits, so it seems a natural solution.

This was discussed in https://github.com/json-ld/yaml-ld/issues/17#issuecomment-1207278736 on #17, and should be the topic of further discussion.

Reply:
Another problem is that the serialization phase comes after compaction and framing, so trying to interpret what nodes represent value objects and are eligible for being serialized using node tags could require recursively processing the internal definition of the graph as it is done in expansion in order to identify which keys represent `@value`, `@type`, and `@language`, so the process of finding those, and replacing them with a scalar with node tag becomes much more difficult.

Consider the following:

```json
{
  "@context": {
    "@vocab": "http://example.org/",
    "prop": {
      "@context": {
        "xsd": "http://www.w3.org/2001/XMLSchema#",
        "val": "@value",
        "dt": "@type"
      }
    }
  },
  "prop": {
    "val": "2022-08-27",
    "dt": "xsd:date"
  }
}
```

The use of an expanded internal representation retains the nature of the literal values so that they are more easily resolved into scalars with node tags.

Of course, when serialized to JSON, or using the YAML-LD JSON schema, these get flattened. The text in this PR simply turns them into strings (or numbers of boolean), but we could decide to turn them into value objects as well, which is also a reasonable direction. At that point, re-compacting might be necessary to get an expression suiting the in-scope context.